### PR TITLE
refactor the slot design of resource group

### PR DESCRIPTION
**Pull request by SmartKeyerror**
After we complete the refactoring of RG memory, I believe it's time to refactor RG slots.

In the current design, slots are used to control the concurrency of a resource group. However, 
the logic is quite complex, and there are many similarities between QD and QE in terms of logic:

1. Slots act as a permit that can be acquired only on QD, and QE doesn't need to concern itself 
with this.
2. Currently, there are several process variables like `self`, `bypassedGroup`, `bypassedSlot`, etc., 
all with similar purposes. These could be merged.
3. During the move query process, inter-process communication is performed using `movetoCallerPid` 
and `movetoResSlot`, which are single fields with multiple layers of meaning. This might lead to 
difficulties in subsequent maintenance.

### The general process

image

1. The `ResGroupSlotData` structure has been completely removed, and `group->nRunning` is 
now used as the virtual slot pool.

2. The process variables `bypassedGroup` and `bypassedSlot` have been eliminated. Only relevant 
information about `self` is retained in QD. For QEs, upon receiving resource group information from 
QD, the `cgroupOpsRoutine->attachcgroup` is directly invoked to add the PID to the Cgroup. No 
information about the currently associated RG is stored in memory.

3. A new field `hasVirtualSlot` has been added to `MyProc`, indicating whether the current process 
has acquired a virtual slot.

### The movq query process

image

The process of move query is more complex compared to the general process, mainly due to the following
reasons:

1. Involves inter-process communication and waiting. The current approach uses signal sending to notify the 
target process and uses Latch to wait for the target process to receive the signal and handle it.
2. Because signals are handled asynchronously, shared memory variables must be handled with caution to 
prevent slot leakage.
3. Handling exceptional situations. For example, signal sending failure, the target process ends before receiving 
the signal, caller process timeouts, and more. We need to handle these cases properly.

To facilitate inter-process communication, I've added the enumeration class `MoveQueryTargetProcessStatus`, 
and its state transitions are as follows:

image

### CODE REVIEW

File: src/backend/postmaster/pgstat.c [hunk 1]
```diff
@@ -4465,6 +4465,65 @@ pgstat_get_backend_current_activity(int pid, bool checkUser)
...
+
+			if (pgstat_read_activity_complete(before_changecount,
+											  after_changecount))
+				break;
+
+			/* Make sure we can break out of loop if stuck... */
+			CHECK_FOR_INTERRUPTS();
+		}
+
+		if (found)
```
Comment (haolinw) [inline src/backend/postmaster/pgstat.c:4518]: Seems better to align with `pgstat_get_backend_current_activity` to have a `checkUser` argument for being as a reminder for other callers (for future development) in which no superuser check before calling this.

**Comment by haolinw** [comment]
Some compiler complains:
```
resgroup.c: In function 'wakeupWaitingVirtualSlotProcs':
resgroup.c:1147:16: error: implicit declaration of function 'groupIsNotDropped'; did you mean 'groupIsDropped'? [-Werror=implicit-function-declaration]
   if (grant && groupIsNotDropped(group))
                ^~~~~~~~~~~~~~~~~
                groupIsDropped
resgroup.c: In function 'resGroupSignalAndWaitTarget':
resgroup.c:2700:16: warning: unused variable 'group' [-Wunused-variable]
  ResGroupData *group;
                ^~~~~
```

**Comment by SmartKeyerror** [comment]
Close this PR temporarily.

**Comment by kainwen** [comment]
> Close this PR temporarily.

I like the PR and the new design. Let's see how to get rid of ABI issue.

**Pull request closed without merging**